monkey/agent_plugins/payloads/cryptojacker/src/cpu_utilizer.py

```diff
@@ -1,12 +1,101 @@
+import hashlib
+import logging
+import threading
+from random import randbytes  # noqa: DUO102 (this isn't for cryptographic use)
+
+import psutil
+
 from common.types import PercentLimited
+from infection_monkey.utils.threading import create_daemon_thread
+
+from .cryptojacker import THREAD_JOIN_TIMEOUT
+
+logger = logging.getLogger(__name__)
+
+
+TARGET_CPU_UTILIZATION = 0.10
+ACCURACY_THRESHOLD = 0.01
+INITIAL_SLEEP_SECONDS = 0.001
+AVERAGE_BLOCK_SIZE_BYTES = int(
+    3.25 * 1024 * 1024
+)  # 3.25 MB - Source: https://trustmachines.co/blog/bitcoin-ordinals-reignited-block-size-debate/
+OPERATION_COUNT = 500
+OPERATION_COUNT_MODIFIER_START = int(OPERATION_COUNT / 10)
+OPERATION_COUNT_MODIFIER_FACTOR = 1.5
+MINIMUM_SLEEP = 0.000001
 
 
 class CPUUtilizer:
     def __init__(self, cpu_utilization: PercentLimited):
-        self._cpu_utilization = cpu_utilization
+        self._should_stop_cpu_utilization = threading.Event()
+        self._cpu_utilizer_thread = create_daemon_thread(
+            target=self._utilize_cpu,
+            name="CPUUtilizationThread",
+            args=(cpu_utilization,),
+        )
 
     def start(self):
-        pass
+        logger.info("Utilizing CPU")
+
+        self._cpu_utilizer_thread.start()
+
+    def _utilize_cpu(self, cpu_utilization: PercentLimited):
+        operation_count_modifier = OPERATION_COUNT_MODIFIER_START
+        sleep_seconds = INITIAL_SLEEP_SECONDS
+        block = randbytes(AVERAGE_BLOCK_SIZE_BYTES)
+        nonce = 0
+
+        process = psutil.Process()
+        # This is a throw-away call. The first call to cpu_percent() always returns 0, even
+        # after generating some hashes.
+        # https://psutil.readthedocs.io/en/latest/#psutil.cpu_percent
+        process.cpu_percent()
+
+        while not self._should_stop_cpu_utilization.is_set():
+            # The operation_count_modifier decreases the number of hashes per iteration.
+            # The modifier, itself, decreases by a factor of 1.5 each iteration, until
+            # it reaches 1. This allows a higher sample rate of the CPU utilization
+            # early on to help the sleep time to converge quicker.
+            for _ in range(0, int(OPERATION_COUNT / operation_count_modifier)):
+                digest = hashlib.sha256()
+                digest.update(nonce.to_bytes(8))
+                digest.update(block)
+                nonce += 1
+
+                self._should_stop_cpu_utilization.wait(sleep_seconds)
+
+            operation_count_modifier = max(  # type: ignore[assignment]
+                operation_count_modifier / OPERATION_COUNT_MODIFIER_FACTOR, 1
+            )
+
+            cpu_utilization = process.cpu_percent() / 100
```

Comment:
Watch your units when you make target utilization configurable. `PercentLimited` objects are 0-100. They do have an `as_decimal_fraction()` method.